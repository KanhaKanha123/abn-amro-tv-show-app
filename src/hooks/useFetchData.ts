import { useEffect, useReducer, useRef, useState } from 'react';
import { formatShowsData } from '../helpers/formatData';
import { InitialState, articleReducer } from '../store/reducers/showsReducer';
import { fetchShows } from '../api/fetchShows';
//custom hook to manage api call to get data
const useFetchData = () => {
    const [state, dispatch] = useReducer(articleReducer, InitialState);
    const page = useRef(0);
    let firstUpdate = useRef<Boolean>(false);

    const getData = async () => {
        try {
            const result = await fetchShows(page.current);

            dispatch({ type: "DATA_SUCCESS", payload: formatShowsData(result) });

            page.current++;

            if (result.length === 0) {
                // hasMoreShows.value = false;
            }
        } catch (e) {
            const err = e as Error;
            dispatch({ type: "DATA_ERROR", payload: err.message });
        } finally {
            // isLoading.value = false;
        }
    };

    useEffect(() => {
        try {
            //react 18 re mount the component, so to call only once maintating by useref
            if (!firstUpdate.current) {
                getData();
            }
            firstUpdate.current = true;
        } catch (err) {
            console.error(err);
        }
    }, []);

    return { state };
}

export default useFetchData;


