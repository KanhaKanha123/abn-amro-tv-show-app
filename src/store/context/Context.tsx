import { createContext, useContext, useReducer } from 'react';
import useFetchData from '../../hooks/useFetchData';
import { filterReducer } from '../reducers/FilterReducer';

type ChildrenProps = {
    children: React.ReactNode
}

const CartContext = createContext<any | null>(null);

const Context = ({ children }: ChildrenProps) => {
    //get state from custome hook
    const { state } = useFetchData();

    //decare reducer for search filter
    const [filterState, filterDispatch] = useReducer(filterReducer, {
        searchTerm: "",
        filteredData: []
    });

    return <CartContext.Provider value={{ state, filterState, filterDispatch }}>{children}</CartContext.Provider>
}

export default Context;

//hook which will provide application wise
export const ShowsAppState = () => {
    return useContext(CartContext);
}
