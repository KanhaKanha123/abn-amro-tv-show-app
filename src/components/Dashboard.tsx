import { Container } from "react-bootstrap";
import useFetchData from "../hooks/useFetchData";
import ShowList from "./show/ShowList";
import './styles/Shows.Style.css';
import { ShowsFilterState } from '../store/context/Context';
import { filterBySearchTerm } from "../helpers/formatData";
import { Fragment } from "react";
import { ModifiedShows } from "../types/types";

const Dashboard = () => {

    const { state: { showsData, error, loading } } = useFetchData();

    const { filterState: { searchTerm } } = ShowsFilterState() || { filterState: { searchTerm: "" } };

    //Filter data here
    const filterData = () => {
        let filteredObj: ModifiedShows[] = showsData;

        if (searchTerm) {
            filteredObj = filterBySearchTerm(showsData, searchTerm);
        }

        return filteredObj;
    };

    return (<div className="dashboard">
        <Container data-testid="dashboard-parent-container" className="parent-container">
            <div className="error-empty-container ">
                {loading && <div data-testid="dashboard-loader" className="loader"></div>}

                {error && <h1 data-testid="dashboard-api-error">{error}</h1>}

                {!error && filterData().length === 0 && <h1 data-testid="dashboard-no-data">No Data Available</h1>}
            </div>
            {(filterData() || []).map((item: ModifiedShows) => (<Fragment key={Math.random() * 1000}>
                {(item.showsList.length > 0 && <div>
                    <span data-testid="dashboard-genre-text" className="genre-text">{item.genreName} (Total Records: {item.showsList.length})</span>
                    <ShowList showsList={item.showsList}></ShowList>
                </div>)}
            </Fragment>))}
        </Container>
    </div>
    )
}

export default Dashboard;