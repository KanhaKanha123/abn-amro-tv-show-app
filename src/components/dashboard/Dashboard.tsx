import { Container } from "react-bootstrap";
import ShowList from "../show/ShowList";
import { ShowsAppState } from '../../store/context/Context';
import { Fragment } from "react";
import { ModifiedShows } from "../../types/types";

const Dashboard = () => {
    //Read state from Context API store
    const { state, filterState: { searchTerm, filteredData } } = ShowsAppState() || ShowsAppState() || { state: { showsData: [], error: "", loading: false }, filterState: { searchTerm: "" }, filterDispatch: () => { } };

    const { showsData, error, loading } = state;

    let filteredShowsData: ModifiedShows[] = [];

    if (searchTerm === "") {
        filteredShowsData = showsData;
    } else {
        filteredShowsData = filteredData;
    }

    return (<Container aria-label="Main tv shows listing container" className="dashboard">
        <main data-testid="dashboard-parent-container" className="parent-container">
            <section aria-label="Section to display error or empty data messages" className="error-empty-container">
                {loading && <div aria-label="Loading spinner" data-testid="dashboard-loader" className="loader"></div>}
                {error && <h1 aria-label="Error message" data-testid="dashboard-api-error">{error}</h1>}
                {!error && filteredShowsData.length === 0 && <h1 aria-label="No tv shows available" data-testid="dashboard-no-data">No Data Available</h1>}
            </section>

            <section aria-label="TV shows listing by genres">
                {filteredShowsData.map((item: ModifiedShows) => (<Fragment key={Math.random() * 1000}>
                    {(item.showsList.length > 0 && <Fragment>
                        <span aria-label="Genre name and total count on top of each table" data-testid="dashboard-genre-text" className="genre-text">{item.genreName} (Total Records: {item.showsList.length})</span>
                        <ShowList showsList={item.showsList}></ShowList>
                    </Fragment>)}
                </Fragment>))}
            </section>
        </main>
    </Container>
    )
}

export default Dashboard;