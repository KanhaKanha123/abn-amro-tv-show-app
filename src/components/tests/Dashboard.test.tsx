import Dashboard from '../Dashboard';
import renderer from 'react-test-renderer';
import { act, render, renderHook, screen, waitFor } from '@testing-library/react';
import useFetchData from '../../hooks/useFetchData';
import { ShowsFilterState } from '../../store/context/Context';

describe('Dashboard component test', () => {

    test("Dashboard component load correctly", () => {

        act(() => { render(<Dashboard />) });
        renderHook(async () => { await act(() => { useFetchData() }) });
        renderHook(async () => { await act(() => { ShowsFilterState() }) });

        waitFor(() => {
            const wrapper = screen.getByTestId("dashboard-parent-container");
            const genreTextContainer = screen.getByTestId("dashboard-genre-text");

            // check element exist
            expect(wrapper).toBeInTheDocument();
            expect(genreTextContainer).toBeInTheDocument();
        });
    });

    test("Dashboard component should match snapshot", async () => {
        const element = await renderer.create(<Dashboard />);
        expect(element.toJSON()).toMatchSnapshot();
    });
});
