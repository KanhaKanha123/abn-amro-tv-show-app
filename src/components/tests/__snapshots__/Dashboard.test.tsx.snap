// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard component test Dashboard component should match snapshot 1`] = `
<div
  className="dashboard"
>
  <div
    className="parent-container container"
    data-testid="dashboard-parent-container"
  >
    <div
      className="error-empty-container "
    >
      <div
        className="loader"
        data-testid="dashboard-loader"
      />
      <h1
        data-testid="dashboard-no-data"
      >
        No Data Available
      </h1>
    </div>
  </div>
</div>
`;
